networkx/algorithms/isomorphism/tree_isomorphism.py

```diff
@@ -1,29 +1,1021 @@
+     151-166.
+
+  .. [4] M. Carrasco-Ruiz, "Herramientas para el Manejo Computacional de
+     Gráficas", Bachelor of Science Thesis, UNAM, (2022): 55-77.
+
 """
 
+from collections import Counter
+
 import networkx as nx
 from networkx.utils.decorators import not_implemented_for
 
 __all__ = ["rooted_tree_isomorphism", "tree_isomorphism"]
 
 
+def categorize_entries(S):
+    """Groups the entries of the lists by its position.
+
+    Given a list of lists whose entries are numbers between 0 and m. Define a
+    NONEMPTY map such that NONEMPTY(i) are the numbers in the i-th position of
+    some list in S; this numbers are sorted.
+
+    For example, given ((1), (1, 2, 3), (1, 2, 2)), NONEMPTY(2) = [2, 3].
+
+    Parameters
+    ----------
+    S : list of lists of ints (natural numbers)
+        The list of lists of natural numbers.
+
+    Returns
+    -------
+    NONEMPTY : dict of int: list of ints
+        The NONEMPTY map such that NONEMPTY(i) are the numbers in the i-th
+        position of some list in S; this list of numbers is sorted.
+
+    Notes
+    -----
+    This algorithm runs in O(l_total + m) time and O(l_total) space, where
+    l_total is the sum of the lengths of the lists in S and m is the max value
+    of all the lists in S.
+
+    """
+    # Define the an empty NONEMPTY map.
+    NONEMPTY = {}
+
+    # The max length found in S.
+    l_max = max([len(s) for s in S])
+
+    # Traverse all the entries contained in the lists of S.
+    for i in range(l_max):
+        # Define a counter to keep track of the entries found in the i-th
+        # position.
+        C = Counter()
+
+        for s in S:
+            # Only consider lists of length greater or equal than i, and whose
+            # i-th entry is not already contained in the multiset.
+            if len(s) < (i + 1) or s[i] in C:
+                continue
+            else:
+                C[s[i]] = 1
+
+        # Obtain a sorted representation of the entries found in the i-th
+        # position.
+        NONEMPTY[i] = [x for x in range(max(C) + 1) for _ in range(C.get(x, 0))]
+
+    return NONEMPTY
+
+
+def categorize_lists(S):
+    """Groups the lists by its length.
+
+    Given a list of lists whose entries are numbers between 0 and m. Define a
+    LENGTH map such that LENGTH(i) are the lists of S with length i.
+
+    For example, given ((1), (1, 2, 3), (1, 2, 2)), NONEMPTY(3) = [(1, 2, 3),
+    (1, 2, 2)].
+
+    Parameters
+    ----------
+    S : list of lists of ints (natural numbers)
+        The list of lists of natural numbers.
+
+    Returns
+    -------
+    LENGTH : dict of int: list of lists of ints
+        The LENGTH map such that LENGTH(i) are the lists of S with length i.
+
+    Notes
+    -----
+    This algorithm runs in O(|S|) time and space.
+
+    """
+    # Define an empty LENGTH map.
+    LENGTH = {}
+
+    # Traverse the lists in S and add them to the LENGTH.
+    for s in S:
+        if len(s) in LENGTH:
```

Comment:
This kind of code (that you also use in other places) can be simplified with a `collections.defaultdict(list)`, if you want.